runtime/common/src/migrations/balances.rs

```diff
@@ -0,0 +1,138 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+use codec::{Decode, Encode, MaxEncodedLen};
+use frame_support::{
+	dispatch::TypeInfo, storage::unhashed, traits::OnRuntimeUpgrade, weights::Weight, RuntimeDebug,
+	StoragePrefixedMap,
+};
+use frame_system::AccountInfo;
+use sp_arithmetic::traits::{EnsureAdd, Zero};
+use sp_runtime::DispatchError;
+use sp_std::prelude::Vec;
+
+/// All balance information for an account.
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Default, RuntimeDebug, MaxEncodedLen, TypeInfo)]
+pub struct OldAccountData<Balance> {
+	/// Non-reserved part of the balance. There may still be restrictions on
+	/// this, but it is the total pool what may in principle be transferred,
+	/// reserved and used for tipping.
+	///
+	/// This is the only balance that matters in terms of most operations on
+	/// tokens. It alone is used to determine the balance when in the contract
+	/// execution environment.
+	pub free: Balance,
+	/// Balance which is reserved and may not be used at all.
+	///
+	/// This can still get slashed, but gets slashed last of all.
+	///
+	/// This balance is a 'reserve' balance that other subsystems use in order
+	/// to set aside tokens that are still 'owned' by the account holder, but
+	/// which are suspendable. This includes named reserve and unnamed reserve.
+	pub reserved: Balance,
+	/// The amount that `free` may not drop below when withdrawing for *anything
+	/// except transaction fee payment*.
+	pub misc_frozen: Balance,
+	/// The amount that `free` may not drop below when withdrawing specifically
+	/// for transaction fee payment.
+	pub fee_frozen: Balance,
+}
+
+pub type OldAccountInfo<Index, Balance> = AccountInfo<Index, OldAccountData<Balance>>;
+
+pub type NewAccountData<Balance> = pallet_balances::AccountData<Balance>;
+
+pub type NewAccountInfo<Index, Balance> = AccountInfo<Index, NewAccountData<Balance>>;
+
+pub struct Migration<T: pallet_balances::Config + frame_system::Config>(
+	sp_std::marker::PhantomData<T>,
+);
+
+impl<T> OnRuntimeUpgrade for Migration<T>
+where
+	T: pallet_balances::Config + frame_system::Config,
+{
+	#[cfg(feature = "try-runtime")]
+	fn pre_upgrade() -> Result<Vec<u8>, DispatchError> {
+		let account_prefix = frame_system::Account::<T>::final_prefix();
+
+		let mut previous_key = account_prefix.to_vec();
+
+		while let Some(next) = sp_io::storage::next_key(&previous_key) {
+			previous_key = next;
+
+			log::info!(
+				"Balances Migration - Processing account key - {}",
+				hex::encode(previous_key.clone())
+			);
+
+			/// The difference between the old and new account data structures
```

Comment:
Tested by running:

```
./target/debug/centrifuge-chain try-runtime --runtime target/debug/wbuild/centrifuge-runtime/centrifuge_runtime.wasm --chain centrifuge on-runtime-upgrade live --uri wss://fullnode.centrifuge.io:443
```

Some entries decode without any issues, however, the majority fail when decoding to both old and new, also with "unexpected" values, as per the checks listed below.

Reply:
After running the above, using the live centrifuge chain, the total number of account keys processed was - 102015.

cc @wischli

Reply:
I suppose that number reflects the decodable acocunts? I'd be interested in the accounts with `reserved` and/or `frozen` balance.

Reply:
We can add those as well, I'm concerned with the accounts that fail to decode into the old structure, I wasn't expecting that tbh.